core/lib/node/src/node/mod.rs

```diff
@@ -0,0 +1,269 @@
+    stop_sender: watch::Sender<bool>,
+    /// Stop receiver to be shared with tasks.
+    stop_receiver: StopReceiver,
+    /// Tokio runtime used to spawn tasks.
+    /// During the node initialization the implicit tokio context is not available, so tasks
+    /// are expected to use the handle provided by [`NodeContext`].
+    runtime: Runtime,
+}
+
+impl fmt::Debug for ZkSyncNode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ZkSyncNode").finish_non_exhaustive()
+    }
+}
+
+impl ZkSyncNode {
+    pub fn new<R: ResourceProvider>(resource_provider: R) -> anyhow::Result<Self> {
+        if tokio::runtime::Handle::try_current().is_ok() {
+            anyhow::bail!(
+                "Detected a Tokio Runtime. ZkSyncNode manages its own runtime and does not support nested runtimes"
+            );
+        }
+        let runtime = tokio::runtime::Builder::new_multi_thread()
+            .enable_all()
+            .build()
+            .unwrap();
+
+        let (stop_sender, stop_receiver) = watch::channel(false);
+        let self_ = Self {
+            resource_provider: Box::new(resource_provider),
+            resources: RefCell::default(),
+            task_constructors: Vec::new(),
+            healthcheck_task_constructor: None,
+            stop_sender,
+            stop_receiver: StopReceiver(stop_receiver),
+            runtime,
+        };
+
+        Ok(self_)
+    }
+
+    /// Adds a task to the node.
+    ///
+    /// The task is not created at this point, instead, the constructor is stored in the node
+    /// and will be invoked during [`ZkSyncNode::run`] method. Any error returned by the constructor
+    /// will prevent the node from starting and will be propagated by the [`ZkSyncNode::run`] method.
+    pub fn add_task<
+        F: FnOnce(&NodeContext<'_>) -> Result<Box<dyn ZkSyncTask>, TaskInitError> + 'static,
+    >(
+        &mut self,
+        name: impl AsRef<str>,
+        task_constructor: F,
+    ) -> &mut Self {
+        self.task_constructors
+            .push((name.as_ref().into(), Box::new(task_constructor)));
+        self
+    }
+
+    /// Returns `true` if the healthcheck task is already set.
+    pub fn has_healthcheck_task(&self) -> bool {
+        self.healthcheck_task_constructor.is_some()
+    }
+
+    /// Adds a healthcheck task to the node.
+    ///
+    /// Healthcheck task is treated as any other task, with the following differences:
+    /// - It is given the list of healthchecks of all the other tasks.
+    /// - Its own healthcheck is ignored (the task is expected to handle its own checks itself).
+    /// - There may be only one healthcheck task per node.
+    ///
+    /// If the healthcheck task is already set, the provided one will override it.
+    pub fn with_healthcheck<
+        F: FnOnce(
+                &NodeContext<'_>,
+                Vec<Box<dyn CheckHealth>>,
+            ) -> Result<Box<dyn ZkSyncTask>, TaskInitError>
+            + 'static,
```

Comment:
Noted on the health checks -- and I think I kind of like this idea, this also allows one to implement other similar concepts, like circuit breakers (probably we want components to define their circuit breakers rather than have to initialize them as a separate entity). What I don't know right now is a good way to do it without having to rely on the invocation order (e.g. even if the healtcheck task is added first, it'll still observe all the healthchecks).

On the sync/async -- the tokio runtime handle is provided through the node context, but I'm still not sure if a freely available async context is a good thing.

Reply:
Everything is async now.